pkg/dns/dns.go

```diff
@@ -0,0 +1,394 @@
+								Address: addr,
+								PortSpecifier: &v3.SocketAddress_PortValue{
+									PortValue: uint32(p.port),
+								},
+							},
+						},
+					},
+				},
+			},
+		}
+		lbEndpoints = append(lbEndpoints, lbEndpoint)
+	}
+
+	if p.cluster.LoadAssignment != nil {
+		p.cluster.LoadAssignment.Endpoints = []*config_endpoint_v3.LocalityLbEndpoints{
+			{
+				LbEndpoints: lbEndpoints,
+			},
+		}
+	}
+}
+
+func NewDNSResolver() (*DNSResolver, error) {
+	r := &DNSResolver{
+		DnsResolverChan: make(chan []*config_cluster_v3.Cluster),
+		cache:           map[string]*domainCacheEntry{},
+		adsCache:        ads.NewAdsCache(),
+		dnsRefreshQueue: workqueue.NewDelayingQueue(),
+		client: &dns.Client{
+			DialTimeout:  5 * time.Second,
+			ReadTimeout:  5 * time.Second,
+			WriteTimeout: 5 * time.Second,
+		},
+	}
+
+	dnsConfig, err := dns.ClientConfigFromFile("/etc/resolv.conf")
+	if err != nil {
+		return nil, err
+	}
+	if dnsConfig != nil {
+		for _, s := range dnsConfig.Servers {
+			r.resolvConfServers = append(r.resolvConfServers, net.JoinHostPort(s, dnsConfig.Port))
+		}
+	}
+
+	return r, nil
+}
+
+func (r *DNSResolver) StartDNSResolver(stopCh <-chan struct{}) {
+	go r.startResolver()
+	go r.refreshDNS()
+	go func() {
+		<-stopCh
+		r.dnsRefreshQueue.ShutDown()
+		close(r.DnsResolverChan)
+	}()
+}
+
+// startResolver watches the DnsResolver Channel
+func (r *DNSResolver) startResolver() {
+	temp := make(chan struct{}, MaxConcurrency)
+	for clusters := range r.DnsResolverChan {
+		temp <- struct{}{}
+
+		go func(clusters []*config_cluster_v3.Cluster) {
+			defer func() { <-temp }()
+			r.resolveDomains(clusters)
+		}(clusters)
+	}
+}
+
+// resolveDomains takes a map of hostnames and refresh rate
+func (r *DNSResolver) resolveDomains(clusters []*config_cluster_v3.Cluster) {
+	domains := getPendingResolveDomain(clusters)
+
+	// Stow domain updates, need to remove unwatched domains first
+	r.removeUnwatchedDomain(domains)
+	for _, v := range domains {
+		r.resolve(v)
+	}
+	r.adsCache.ClusterCache.Flush()
+}
+
+// removeUnwatchedDomain cancels any scheduled re-resolve for names we no longer care about
+func (r *DNSResolver) removeUnwatchedDomain(domains map[string]*pendingResolveDomain) {
+	r.Lock()
+	defer r.Unlock()
+	for name, value := range r.cache {
+		if _, ok := domains[name]; ok {
+			continue
+		}
+		r.adsCache.UpdateApiClusterStatus(value.clusterName, core_v2.ApiStatus_DELETE)
+		delete(r.cache, name)
+	}
+}
+
+// This functions were copied and adapted from github.com/istio/istio/pilot/pkg/model/network.go.
+func (r *DNSResolver) resolve(v *pendingResolveDomain) {
+	if entry, ok := r.cache[v.domainName]; ok && entry.expiry.After(time.Now()) {
+		return
+	}
+
+	r.Lock()
+	// ideally this will not happen more than once for each name and the cache auto-updates in the background
+	// even if it does, this happens on the SotW ingestion path (kube or meshnetworks changes)
+	entry := r.cache[v.domainName]
+	delete(r.cache, v.domainName)
```

Comment:
Assume this is checking for dns refresh not for notification from ads_processor, can you move it out of the function to the periodic refresher

Reply:
I am also not sure why it is delted and resolve again